### PR TITLE
Service Locator mark I. Service Locator and Analytics

### LINKED ISSUE #1197: Global dependencies: Service Locator and Analytics
Following the discussion in p99K0U-1BK-p2, we are going to abstract the global dependencies behind a Service Locator.

The scope for this ticket is:
- [ ] Add the Service Locator
- [ ] Provide access to WooAnalytics through the service locator. WooAnalytics is not a singleton anymore
- [ ] Integrate the change in call sites to WooAnalytics

### PR DESCRIPTION
**Pull request by ctarda**
Fixes #1197 

As discussed in p99K0U-1BK-p2 we are going to start fetching the global dependencies from a [Service Locator](https://en.wikipedia.org/wiki/Service_locator_pattern). 

The refactoring will be split in several PRs., this is the first one. The scope for this first PR is:
- [ ] Add the Service Locator
- [ ] Provide access to WooAnalytics through the service locator. WooAnalytics is not a singleton anymore
- [ ] Integrate the change in call sites to WooAnalytics

## Changes
- Add Service Locator and tests
- Add an interface to abstract WooAnalytics, called Analytics
- Move `MockupAnalyticsProvider` from the main target back to the testing target, undoing the change in PR #1130 

## Testing
- Checkout the branch, build the project and run the tests.
- Make sure that tests pass
- Make sure that Tracks events are still being submitted. It wouldn't hurt to check in the actual Tracks that the events are being received.

### CODE REVIEW

File: WooCommerce/Classes/ServiceLocator/Analytics.swift [hunk 1]
```diff
@@ -0,0 +1,49 @@
+import Foundation
+
+/// Abstracts the Analytics engine.
+///
+protocol Analytics {
+    /// Initialize the analytics engine
+    ///
+    func initialize()
+
+    /// Track a spcific event without any associated properties
```
Comment (mindgraffiti) [inline WooCommerce/Classes/ServiceLocator/Analytics.swift]:
```suggestion
    /// Track a specific event without any associated properties
```

File: WooCommerce/Classes/ServiceLocator/ServiceLocator.swift [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import Foundation
+
+
+/// Provides global depedencies.
+///
+final class ServiceLocator {
+    private static var _analytics: Analytics = WooAnalytics(analyticsProvider: TracksProvider())
+
+    /// Provides the access point to the analytics.
+    /// - Returns: An iumplementation of the Analytics protocol. It defaults to WooAnalytics
```
Comment (mindgraffiti) [inline WooCommerce/Classes/ServiceLocator/ServiceLocator.swift]:
```suggestion
    /// - Returns: An implementation of the Analytics protocol. It defaults to WooAnalytics
```

File: WooCommerce/Classes/ServiceLocator/Analytics.swift [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    func refreshUserData()
+
+    /// Sets the boolean flag to signal that the user has opted out of analytics
+    /// It will trigger a call to `refreshUserData` before starting tracking again.
+    ///
+    func setUserHasOptedOut(_ optedOut: Bool)
+
+    /// Check user opt-in for analytics
+    ///
+    var userHasOptedIn: Bool { get set }
```
[the remark this answers is not in the document]
Comment (mindgraffiti) [reply WooCommerce/Classes/ServiceLocator/Analytics.swift]: I see where you're coming from. Even if this was an optimal change, it doesn't apply to the scope of this PR. The checks for opt in / opt out are complicated and could be optimized a bit. Would you file a `tech debt` issue for it?

File: WooCommerce/Classes/ServiceLocator/ServiceLocator.swift [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+        return _analytics
+    }
+}
+
+
+// MARK: - Testability
+
+/// The setters declared in this extension are meant to be used only from the test bundle
+extension ServiceLocator {
+    static func setAnalytics(_ mock: Analytics) {
```
[the remark this answers is not in the document]
Comment (ctarda) [reply WooCommerce/Classes/ServiceLocator/ServiceLocator.swift]: Nice, TIL!

**Pull request merged**